consensus/hotstuff/timeoutcollector/timeout_processor.go

```diff
@@ -0,0 +1,300 @@
+// Concurrency safe.
+type TimeoutProcessor struct {
+	view             uint64
+	validator        hotstuff.Validator
+	committee        hotstuff.Replicas
+	sigAggregator    hotstuff.TimeoutSignatureAggregator
+	notifier         hotstuff.TimeoutCollectorConsumer
+	partialTCTracker accumulatedWeightTracker
+	tcTracker        accumulatedWeightTracker
+	newestQCTracker  *NewestQCTracker
+}
+
+var _ hotstuff.TimeoutProcessor = (*TimeoutProcessor)(nil)
+
+// NewTimeoutProcessor creates new instance of TimeoutProcessor
+// Returns the following expected errors for invalid inputs:
+//   * model.ErrViewForUnknownEpoch if no epoch containing the given view is known
+// All other errors should be treated as exceptions.
+func NewTimeoutProcessor(committee hotstuff.Replicas,
+	validator hotstuff.Validator,
+	sigAggregator hotstuff.TimeoutSignatureAggregator,
+	notifier hotstuff.TimeoutCollectorConsumer,
+) (*TimeoutProcessor, error) {
+	view := sigAggregator.View()
+	qcThreshold, err := committee.QuorumThresholdForView(view)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve QC weight threshold for view %d: %w", view, err)
+	}
+	timeoutThreshold, err := committee.TimeoutThresholdForView(view)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve timeout weight threshold for view %d: %w", view, err)
+	}
+	return &TimeoutProcessor{
+		view:      view,
+		committee: committee,
+		validator: validator,
+		notifier:  notifier,
+		partialTCTracker: accumulatedWeightTracker{
+			minRequiredWeight: timeoutThreshold,
+			done:              *atomic.NewBool(false),
+		},
+		tcTracker: accumulatedWeightTracker{
+			minRequiredWeight: qcThreshold,
+			done:              *atomic.NewBool(false),
+		},
+		sigAggregator:   sigAggregator,
+		newestQCTracker: NewNewestQCTracker(),
+	}, nil
+}
+
+// Process performs processing of timeout object in concurrent safe way. This
+// function is implemented to be called by multiple goroutines at the same time.
+// Design of this function is event driven, as soon as we collect enough weight
+// to create a TC or a partial TC we will immediately do so and submit it
+// via callback for further processing.
+// Expected error returns during normal operations:
+// * ErrTimeoutForIncompatibleView - submitted timeout for incompatible view
+// * model.InvalidTimeoutError - submitted invalid timeout(invalid structure or invalid signature)
+// * model.ErrViewForUnknownEpoch if no epoch containing the given view is known
+// All other errors should be treated as exceptions.
+func (p *TimeoutProcessor) Process(timeout *model.TimeoutObject) error {
+	if p.view != timeout.View {
+		return fmt.Errorf("received incompatible timeout, expected %d got %d: %w", p.view, timeout.View, ErrTimeoutForIncompatibleView)
+	}
+
+	if p.tcTracker.Done() {
+		return nil
+	}
+
+	err := p.validateTimeout(timeout)
+	if err != nil {
+		return fmt.Errorf("received invalid timeout: %w", err)
+	}
+
+	p.newestQCTracker.Track(timeout.NewestQC)
```

Comment:
Lets add some documentation to explain that it is critical for correctness that `newestQCTracker` is updated _before_ `sigAggregator`

```suggestion
	// CAUTION: for correctness it is critical that we update the `newestQCTracker` first, _before_ we add the
	// TO's signature to `sigAggregator`. Reasoning:
	//  * For a valid TC, we require that the TC includes a QC with view ≥ max{TC.NewestQCViews}.
	//  * The `NewestQCViews` is maintained by `sigAggregator`.
	//  * Hence, for any view `v ∈ NewestQCViews` that `sigAggregator` knows, a QC with equal or larger view is
	//    known to `newestQCTracker`. This is guaranteed if and only if `newestQCTracker` is updated first. 
	p.newestQCTracker.Track(timeout.NewestQC)
```